Add LOCATION option to VORON_PURGE.

cdheiser opened the pull request:
Permissible values are:
  - LEFT or FRONT_LEFT (the default and historical location)
  - RIGHT or FRONT_RIGHT (to the right front of the print)
  - REAR_LEFT
  - REAR_RIGHT

New variables have been added to KAMP_Settings.cfg to tell the macro the max X and Y locations for a purge.

This is incredibly useful for IDEX and multi-toolhead printers as each extruder can use a different location to purge.

cdheiser committed:
Add LOCATION option to VORON_PURGE.

Permissible values are:
  - LEFT or FRONT_LEFT (the default and historical location)
  - RIGHT or FRONT_RIGHT (to the right front of the print)
  - REAR_LEFT
  - REAR_RIGHT

New variables have been added to KAMP_Settings.cfg to tell the macro the max X and Y locations for a purge.

This is incredibly useful for IDEX and multi-toolhead printers as each extruder can use a different location to purge.

kyleisah commented:
This sounds kind of neat, but something I'm worried about is it may cause some confusion for people who have it currently installed. Updating via moonraker doesn't update the KAMP_Settings file outside of the KAMP folder so they may not even be aware of the feature. 

It also breaks if they're aware of the feature, but haven't updated the KAMP settings file. 

I think max_x and max_y can be pulled from the printer config file. `printer.configfile.settings.stepper_x.position_max` or something like that.

cdheiser commented:
On Fri, Feb 23, 2024 at 4:17 AM kyleisah ***@***.***> wrote:

> This sounds kind of neat, but something I'm worried about is it may cause
> some confusion for people who have it currently installed. Updating via
> moonraker doesn't update the KAMP_Settings file outside of the KAMP folder
> so they may not even be aware of the feature.
>
Not sure how to address this.  It's somewhat consistent with how most
klipper add-ons I've used are (and klipper itself for that matter).

> It also breaks if they're aware of the feature, but haven't updated the
> KAMP settings file.
>
I can address this by preventing usage of the feature without max_x and
max_y defined.

> I think max_x and max_y can be pulled from the printer config file.
> printer.configfile.settings.stepper_x.position_max or something like that.
>
You could, but you have the following problems:

1) You could end up trying to purge off the bed.  (Printers can and do have
X and Y travel that exceeds the bed size

2) In IDEX printers, the max X is different depending on the carriage.

> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kyleisah/Klipper-Adaptive-Meshing-Purging/pull/231#issuecomment-1961225805>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQARWUDXSDLAIUK46P6YYDYVCCEJAVCNFSM6AAAAABBPGSYS6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRRGIZDKOBQGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

cdheiser committed:
Make the LOCATION option for VORON_PURGE more user friendly:

- Defaults for max_x and max_y are set to -1, preventing accidental usage with incorrect values
- Macro now also recognizes when a user has upgraded KAMP but failed to add the new config options

cdheiser commented:
I've updated the PR with tweaks so users MUST set proper values for max_x and max_y and the macro has defaults that prevent the use of a location without the appropriate configuration variable defined.

santiagomoneta commented:
I noticed that the location of the purge don't take the brim into account and I see it tries to do the brim ON TOP of the purge line

cdheiser commented:
That issue doesn't really have anything to do with this PR.  That's
generally a slicer/configuration problem.  This is why KAMP allows you to
configure the distance to the objects for the purge because the slicer
doesn't include this information in the metadata.  Furthermore, the purge
tower is also not included and thus it's possible to have the purge tower
print on top of the purge lines.

On Sat, Apr 20, 2024, 6:01 AM Santiago Moneta ***@***.***>
wrote:

> I noticed that the location of the purge don't take the brim into account
> and I see it tries to do the brim ON TOP of the purge line
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kyleisah/Klipper-Adaptive-Meshing-Purging/pull/231#issuecomment-2067666479>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQARWV3OEUY2WGLURRRRUTY6JRKPAVCNFSM6AAAAABBPGSYS6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANRXGY3DMNBXHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

cdheiser commented:
Bump... any more thoughts on this?  It's been working great and also doesn't cause any issues for existing users that update.

kyleisah commented:
> Bump... any more thoughts on this?  It's been working great and also doesn't cause any issues for existing users that update.  

Hey there, appreciate you checking in. Unfortunately the repo is currently on a bit of a hold until I figure out a direction to go from here. Adaptive meshing is in mainline klipper now, and the adaptive purging is still well loved, which I why I don't want to touch anything until I have a solid plan. I appreciate the work you've put in, don't get me wrong. 

I'm currently noodling with some refactoring of the adaptive purging. Something that will really cook. I have big plans, it's just gonna take some work and time.

I've been working on learning python for scripting, rather than be stuck with klipper macros. 

When I get something that is intuitive enough and packaged neatly, I'll be ready to update the repo. Are you by any chance on Discord? Feel free to DM me, my discord user ID is: `704121656460705944`. Would be happy to share what I've been working on there.

kyleisah commented:
Actually, I guess it shouldn't be a secret. Here's what I'm working on:
![Slicer Handled Voron Purge](https://github.com/kyleisah/Klipper-Adaptive-Meshing-Purging/assets/40312912/1c341717-bf5a-4053-8193-057baf2675d5)
The gif is terrbile quality, but I've been working on a python script that will take some variables you enter (much like KAMP_Settings.cfg) and translates it into slicer gcode so you can visualize the purge right on your buildplate. It actually works even better than KAMP's current purging. No more brim/skirt collision, purge towers are accounted for, etc. It's amazing.

MHillyer commented:
Also voting to add this. I installed a conveyor belt on my printer, so I need my purge to run along the Y axis instead of along the X axis (if it is on the X axis it will just roll right along the front of the belt and under the build area, where if it's on the Y axis it peels off as the belt turns).

I manually edited the purge code to reverse the default check for room along Y then printing along X to be check for room along X and print down the Y but now I have a dirty repo warning, would rather have something officially supported.
